frontends/p4/substitution.h

```diff
@@ -76,21 +76,11 @@ class TypeVariableSubstitution final : public TypeSubstitution<const IR::ITypeVa
  public:
     TypeVariableSubstitution() = default;
     TypeVariableSubstitution(const TypeVariableSubstitution& other) = default;
-    /* Set bindings for multiple variables at the same time.
-     * @param params   List of variables to set.
-     * @param args     List of type arguments to set to variables.
-     * @param root     P4 program element which is being processed. Used for error reporting.
-     * @return         True on success. */
-    bool setBindings(const IR::Node* root,
```

Comment:
Turning it into a doxygen style comment is just a matter of starting the comment with '/**'. Doxygen supports most of the JavaDoc annotations.

Indeed, having one person document all the APIs in one shot will be daunting. However, documenting APIs as people touch them, would spread around the pain and significantly improve our ability to help people on-board on the compiler faster. Given that we want the compiler to be adopted by a large set of developers for a variety of targets, this is quite desirable.

Beside API comments, I discussed with @ChrisDodd documenting what the passes do and why, what are the invariants for each pass, and what are the assumptions on other passes. These are things that are hard to glean just by looking at the code, and again, it will really help get other developers to contribute to the compiler.

I am not familiar with your development environment, however, I thought Visual Studio provides good support for C++. If you develop in Linux, there are a number of IDE options with good C++ support as well: Qt Creator (http://wiki.qt.io/Main) and Eclipse CDT (https://eclipse.org/cdt/) have made good progress in handling largish projects. They also integrate with git to alleviate some of your git pains :)